src/ParticleDA.jl

```diff
@@ -362,29 +362,46 @@ function init_filter(filter_params::FilterParameters, model_data, nprt_per_rank:
     online_matrices = init_online_matrices(grid, grid_ext, stations, filter_params, T)
     rng = get_rng(model_data)
 
-    # Returning data as a named tuple to simplify the calling code. The Julia 1.5 syntax would be nice here.
-    return (filter_data = filter_data,
-            offline_matrices = offline_matrices,
-            online_matrices = online_matrices,
-            stations = stations,
-            grid = grid,
-            grid_ext = grid_ext,
-            rng = rng,
-            obs_noise_std = obs_noise_std)
+    return (; filter_data..., offline_matrices, online_matrices, stations, grid, grid_ext, rng, obs_noise_std)
 end
 
-struct FilterData{T, S, U, V, X}
+function update_particle_proposal!(model_data, filter_data, filter_params, truth_observations, nprt_per_rank, filter_type::BootstrapFilter)
 
-    weights::T
-    resampling_indices::U
-    statistics::S
-    avg_arr::V
-    var_arr::V
-    copy_buffer::X
+    update_particle_noise!(model_data, nprt_per_rank)
 
 end
 
-function run_particle_filter(init, filter_params::FilterParameters, model_params_dict::Dict, ::BootstrapFilter)
+function update_particle_proposal!(model_data, filter_data, filter_params, truth_observations, nprt_per_rank, filter_type::OptimalFilter)
+
+        # Optimal Filter: After updating the particle dynamics, we apply the "optimal proposal" in
+        #                 sample_height_proposal!() to the first state variable (height). We apply
+        #                 a sample from the gaussian random field in update_particle_noise!() to the other
+        #                 state variables (velocity).
+
+        particles = get_particles(model_data)
+
+        # Apply optimal proposal, the result will be in offline_matrices.samples
+        sample_height_proposal!(@view(particles[:,:,1,:]),
+                                filter_data.offline_matrices,
+                                filter_data.online_matrices,
+                                truth_observations,
+                                filter_data.stations,
+                                filter_data.grid,
+                                filter_data.grid_ext,
+                                filter_params,
+                                filter_data.rng[threadid()],
+                                filter_data.obs_noise_std)
+
+        # Add noise from the standard gaussian random field to all state variables in model_data
+        update_particle_noise!(model_data, nprt_per_rank)
+
+        # Overwrite the height state variable with the samples of the optimal proposal
+        particles[:,:,1,:] .= filter_data.online_matrices.samples
+        set_particles!(model_data, particles)
+    
+end
+
+function run_particle_filter(init, filter_params::FilterParameters, model_params_dict::Dict, filter_type)
```

Comment:
I'd annotate the last argument with the type
```suggestion
function run_particle_filter(init, filter_params::FilterParameters, model_params_dict::Dict, filter_type::ParticleFilter)
```